### PR TITLE
Multiple updates to ESIGMA IMR model

### PR DESCRIPTION
**Pull request by prayush**
### Multiple updates to ESIGMA IMR model:

 - Add option to use orbit averaged frequency to determine hybridization frequency window
 - Add global constants to represent ESIGMA eccleanups
 - use delta_t consistently
- Reorganize code to improve readability
- Update ESIGMA hybridization:
   - Allow user to choose either the initial orbital mean anomaly or coalescence phase to specify
   - Update hybridization code to allow for phase aligning of inspiral to merger,
   and vice versa

Putting this up for discussion @Akash-Maurya-0899, @Madd-Scientist  and @Kaushik

### CODE REVIEW

**Comment by prayush** [comment]
@Akash-Maurya-0899 awaiting your comments

File: gwnr/waveform/esigma_utils.py [hunk 1]
```diff
@@ -443,23 +446,190 @@ def get_inspiral_esigma_waveform(
...
+            window_end_idx = np.searchsorted(
+                orbital_phase, orbital_phase[transition_idx] + num_hyb_orbits * np.pi
+            )
+        f_window_mr_transition = 4 * np.max(
+            [
+                orbital_freq[window_end_idx] - f_mr_transition / 2.0,
+                f_mr_transition / 2.0 - orbital_freq[window_start_idx],
+            ]
+        )  # Extra 2-factor for returning the (2,2)-mode frequency
+
```
Comment (Akash-Maurya-0899) [inline gwnr/waveform/esigma_utils.py:606]: Simplification of the above conditions:

```
    if keep_f_mr_transition_at_center:
        # Orbital cycle based hybridization that keeps f_mr_transition at
        # the hybridization frequency window's midpoint
        if (orbital_phase[transition_idx] + num_hyb_orbits * np.pi) > orbital_phase[-1]:
            raise RuntimeError(
                f"""Requested number of orbits to hybridize over not available
in the waveform after the transition frequency. Either decrease the number of
orbits to hybridize over (currently {num_hyb_orbits}) or decrease the
inspiral-to-merger transition frequency."""
            )

        if hybridize_using_orbital_frequency:
            window_start_idx = _get_window_start(
                orbital_freq[: transition_idx + 1],
                delta_t,
                num_hyb_orbits * np.pi,
                direction="backward",
            )
            window_end_idx = transition_idx + _get_window_start(
                orbital_freq[transition_idx:],
                delta_t,
                num_hyb_orbits * np.pi,
                direction="forward",
            )
        else:
            # phase is always a monotonically increasing function,
            #    hence using the more efficient np.searchsorted method
            window_start_idx = -1 + np.searchsorted(
                orbital_phase, orbital_phase[transition_idx] - num_hyb_orbits * np.pi
            )
            window_end_idx = np.searchsorted(
                orbital_phase, orbital_phase[transition_idx] + num_hyb_orbits * np.pi
            )
        f_window_mr_transition = 4 * np.max(
            [
                orbital_freq[window_end_idx] - f_mr_transition / 2.0,
                f_mr_transition / 2.0 - orbital_freq[window_start_idx],
            ]
        )  # Extra 2-factor for returning the (2,2)-mode frequency
   
    else:
        if hybridize_using_orbital_frequency:
            window_start_idx = _get_window_start(
                orbital_freq[: transition_idx + 1],
                delta_t,
                2 * num_hyb_orbits * np.pi,
                direction="backward",
            )
        else:
            # Orbital cycle based hybridization that keeps f_mr_transition
            # at the hybridization frequency window's end
            window_start_idx = (
                np.searchsorted(
                    orbital_phase,
                    orbital_phase[transition_idx] - 2 * np.pi * num_hyb_orbits,
                )
                - 1
            )
        f_window_mr_transition = 2 * (
            orbital_freq[transition_idx] - orbital_freq[window_start_idx]
        )  # Extra 2-factor for returning the (2,2)-mode frequency

        if verbose > 4:
            print(
                f"""The transition is to happen in a window from
                frequencies: [{orbital_freq[window_start_idx]}, {orbital_freq[transition_idx]}]Hz,
                between indices: [{window_start_idx}, {transition_idx}]"""
            )
```

This avoids `failsafe` altogether (hence it can be removed from this function), saves some unnecessary checks of the condition `orbital_phase[transition_idx] + num_hyb_orbits * np.pi) > orbital_phase[-1]` and saves the unnecessary hybridization window computation for the case when `keep_f_mr_transition_at_center`, `failsafe` and `orbital_phase[transition_idx] + num_hyb_orbits * np.pi) > orbital_phase[-1]` are all True, in which case the RunTimeError will be executed anyway. 

Also regarding this last condition, it means that even if `failsafe=True` and  `orbital_phase[transition_idx] + num_hyb_orbits * np.pi) > orbital_phase[-1]`, and if the user demands `keep_f_mr_transition_at_center`, we are not switching to the scheme where we keep the transition frequency at the edge. It seems fine to me as switching from centre to edge would not be good for parameter-sapce continuity of the waveform as well. But just highlighting this case for your comments

File: gwnr/waveform/esigma_utils.py [hunk 2]
```diff
@@ -579,132 +793,98 @@ def get_imr_esigma_modes(
...
-            print(f"""f_window_mr_transition = {f_window_mr_transition}""")
+        f_window_mr_transition = _get_transition_frequency_window(
+            retval[-2]["phi"],
+            orbital_frequency,
+            delta_t,
+            f_mr_transition,
+            num_hyb_orbits,
+            keep_f_mr_transition_at_center,
+            hybridize_using_orbital_frequency,
+            failsafe=True,
```
Comment (Akash-Maurya-0899) [inline gwnr/waveform/esigma_utils.py:885]: `failsafe` can be removed here I guess (see my comment on the code of `_get_transition_frequency_window`)

**Review by Akash-Maurya-0899 — CHANGES_REQUESTED** [review]

File: gwnr/waveform/esigma_utils.py
[the remark this answers is not in the document]
Comment (prayush) [reply gwnr/waveform/esigma_utils.py:1]: fixed

File: gwnr/waveform/esigma_utils.py [hunk 3]
```diff
@@ -443,23 +446,190 @@ def get_inspiral_esigma_waveform(
...
+    """
+    from scipy import integrate
+
+    if direction == "backward":
+        for idx in range(len(freq_) - 2, 0, -1):
+            # this could be optimized to not repeat integration
+            if abs(integrate.trapezoid(freq_[idx:], dx=delta_t_)) >= delta_phi_:
+                return idx
+    elif direction == "forward":
+        for idx in range(1, len(freq_) - 1):
```
[the remark this answers is not in the document]
Comment (prayush) [reply gwnr/waveform/esigma_utils.py:471]: changed

File: gwnr/waveform/esigma_utils.py [hunk 4]
```diff
@@ -443,23 +446,190 @@ def get_inspiral_esigma_waveform(
...
+    """
+    This function first finds where the `orbital_freq` crosses the transition
+    frequency `f_mr_transition`, and finds the point in time that is
+    `num_hyb_orbits` orbits before that time. This marks the start of the
+    hybridization window, and the orbital frequency at that time is returned
+
+    Parameters:
+    -----------
+        orbital_phase             -- Luminosity distance to the binary (in Mpc)
+        orbital_freq              -- GW modes to use. List of tuples (l, |m|)
```
[the remark this answers is not in the document]
Comment (prayush) [reply gwnr/waveform/esigma_utils.py:497]: fixed

File: gwnr/waveform/esigma_utils.py [hunk 5]
```diff
@@ -443,23 +446,190 @@ def get_inspiral_esigma_waveform(
...
+                                     hybridize modes, instead of the modes'
+                                     frequency.
+        failsafe                  -- If True, we make reasonable choices for the
+                                     user, if the inputs to this method lead
+                                     into exceptions.
+        verbose                   -- Verbosity flag
+
+    Returns:
+    --------
+        f_window_mr_transition    -- Dictionary of IMR GW modes PyCBC TimeSeries
```
[the remark this answers is not in the document]
Comment (prayush) [reply gwnr/waveform/esigma_utils.py:519]: fixed

File: gwnr/waveform/esigma_utils.py [hunk 6]
```diff
@@ -579,132 +793,98 @@ def get_imr_esigma_modes(
...
-            if verbose > 4:
-                print(
-                    f"""The transition is to happen in a window from
-                    frequencies: [{orb_freq[window_start_idx]}, {orb_freq[transition_idx]}]Hz,
-                    between indices: [{window_start_idx}, {transition_idx}]"""
-                )
-        if verbose > 4:
-            print(f"""f_window_mr_transition = {f_window_mr_transition}""")
+        f_window_mr_transition = _get_transition_frequency_window(
+            retval[-2]["phi"],
```
[the remark this answers is not in the document]
Comment (prayush) [reply gwnr/waveform/esigma_utils.py:878]: replaced with

```
return_orbital_params = return_orbital_params.union(set(["x", "phi", "phidot"]))
```

No need to be too careful with what is kept, these quantities have already been computed and that is the expensive part.

File: gwnr/waveform/esigma_utils.py [hunk 7]
```diff
@@ -752,20 +932,23 @@ def get_imr_esigma_modes(
...
     try:
         retval = gwnr.waveform.hybridize.hybridize_modes(
-            modes_numpy,
+            modes_inspiral_numpy,
             modes_mr_numpy,
+            orbital_frequency,
             f_mr_transition,
             frq_width=f_window_mr_transition,
-            delta_t=1.0 / sample_rate,
+            delta_t=delta_t,
```
[the remark this answers is not in the document]
Comment (prayush) [reply gwnr/waveform/esigma_utils.py:940]: We will keep `delta_t` everywhere, and not use `sample_rate` rounded to the nearest integer. We will ask the user to keep `delta_t = 1 / power_of_2` or face consequent slowdowns in FFT etc.

File: gwnr/waveform/esigma_utils.py [hunk 8]
```diff
@@ -781,7 +964,7 @@ def get_imr_esigma_modes(
     modes_imr = {}
     for el, em in modes_imr_numpy:
         modes_imr[(el, em)] = pt.TimeSeries(
-            modes_imr_numpy[(el, em)], delta_t=1.0 / sample_rate, epoch=-1 * t_peak
+            modes_imr_numpy[(el, em)], delta_t=delta_t, epoch=-1 * t_peak
```
[the remark this answers is not in the document]
Comment (prayush) [reply gwnr/waveform/esigma_utils.py:967]: same response

File: gwnr/waveform/esigma_utils.py [hunk 9]
```diff
@@ -542,8 +755,9 @@ def get_imr_esigma_modes(
         f_mr_transition = min(f_Kerr, f_Schwarz)
```
[the remark this answers is not in the document]
Comment (prayush) [reply gwnr/waveform/esigma_utils.py:755]: changed

File: gwnr/waveform/esigma_utils.py [hunk 10]
```diff
@@ -579,132 +793,98 @@ def get_imr_esigma_modes(
...
+        mode_frequency = gwnr.waveform.hybridize.compute_frequency(mode_phase, delta_t)
         print(
-            f"""DEBUG: Orbital freq at end of inspiral is {orb_freq[-1]}Hz,
-                mode-22 freq at the end of inspiral is {mode_frq[-1]}Hz,
-                max and min mode-22 frequencies are {np.max(mode_frq)}Hz and {np.min(mode_frq)}Hz,
-                and the transition frequency (of {el},{em}-mode) requested is
-                {f_mr_transition}Hz, which should be less than 2 x end-of-inspiral orbital freq {2.0 * orb_freq[-1]}Hz."""
+            f"""DEBUG: Orbital freq at end of inspiral is {orbital_frequency[-1]}Hz,
+mode-22 freq at the end of inspiral is {mode_frequency[-1]}Hz, max and min mode-22
+frequencies are {np.max(mode_frequency)}Hz and {np.min(mode_frequency)}Hz, and the
```
[the remark this answers is not in the document]
Comment (prayush) [reply gwnr/waveform/esigma_utils.py:843]: done

File: gwnr/waveform/esigma_utils.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Akash-Maurya-0899) [reply gwnr/waveform/esigma_utils.py:942]: In that case, please remove `sample_rate` from the code altogether

File: gwnr/waveform/esigma_utils.py
[the remark this answers is not in the document]
Comment (prayush) [reply gwnr/waveform/esigma_utils.py:1]: Good catch. But I think the hybridization function is fine as is ,both `frq_attach` and `frq_width` need to correspond to the mode specified in `mode_to_align_by`, which is why I divided this frequency by the `em` value of `mode_to_align_by` when calculating `t2_index_insp`. The calculation of `t{1,2}_index_mr` is also fine assuming the same definition. What I believe needs to be cared for is the `get_imr_esigma_modes()`. In there we need to be treating `f_mr_transition` and `f_window_mr_transition` as being the frequencies of mode `mode_to_align_by`

[the remark this answers is not in the document]
Comment (prayush) [reply gwnr/waveform/esigma_utils.py:1]: Just pushed a fix ensuring the last part.

File: gwnr/waveform/esigma_utils.py [hunk 11]
```diff
@@ -542,8 +755,9 @@ def get_imr_esigma_modes(
         f_mr_transition = min(f_Kerr, f_Schwarz)
 
     if not return_orbital_params:
-        return_orbital_params = ["e"]
         # Needed necessarily for hybridization error printing
+        return_orbital_params = ["e"]
+
     return_orbital_params = set(return_orbital_params)
```
[the remark this answers is not in the document]
Comment (prayush) [reply gwnr/waveform/esigma_utils.py:774]: fixed

File: gwnr/waveform/esigma_utils.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (prayush) [reply gwnr/waveform/esigma_utils.py:960]: done

File: gwnr/waveform/esigma_utils.py (hunk 1 above)
Comment (prayush) [reply gwnr/waveform/esigma_utils.py:604]: I am not sure I understand this, this seems to me the same exact code logic with the if-else cases reversed in order + the case for `failsafe=True` eliminated entirely. 

Lets discuss this in person?

Comment (prayush) [reply gwnr/waveform/esigma_utils.py:604]: If `failsafe` is activated, it assumes we want to now keep the alignment window's `f_mr_transition` at the end of the window, i.e. it automatically disables `keep_f_mr_transition_at_center`. This is because `f_mr_transition` is set to `mode_frequency.max()`, which could very well be at the last point of inspiral frequency evolution.

Comment (prayush) [reply gwnr/waveform/esigma_utils.py:604]: Remember that if `failsafe` is activated because `mode_frequency.max() < f_mr_transition` then the model will not be continuous across the parameter space anyway. We have literally no way of avoiding that.

Comment (prayush) [reply gwnr/waveform/esigma_utils.py:604]: And the code above automatically activates the `not keep_f_mr_transition_at_center` in `failsafe` mode when its unable to keep that frequency at center since `(orbital_phase[transition_idx] + num_hyb_orbits * np.pi) > orbital_phase[-1]`.

This seems like reasonable `failsafe` behavior. We can also shrink the `f_mr_transition` centered window, that is another option.

File: gwnr/waveform/esigma_utils.py
[the remark this answers is not in the document]
Comment (Akash-Maurya-0899) [reply gwnr/waveform/esigma_utils.py:1]: Yes, exactly

File: gwnr/waveform/esigma_utils.py (hunk 1 above)
Comment (prayush) [reply gwnr/waveform/esigma_utils.py:643]: Just pushed a fixup for this @Akash-Maurya-0899 , as we discussed today.

File: gwnr/waveform/esigma_utils.py (hunk 2 above)
Comment (prayush) [reply gwnr/waveform/esigma_utils.py]: fixed

File: gwnr/waveform/esigma_utils.py [hunk 12]
```diff
@@ -468,83 +675,132 @@ def get_imr_esigma_modes(
...
-    if return_orbital_params is True:
+    if return_orbital_params == True:
         return_orbital_params = ["x", "e", "l", "phi", "phidot", "r", "rdot"]
-
-    if isinstance(return_orbital_params, list):
+    elif isinstance(return_orbital_params, list):
         return_orbital_params_user = return_orbital_params.copy()
-    else:
+        return_orbital_params_user = set(return_orbital_params_user).intersection(
+            set(["x", "e", "l", "phi", "phidot", "r", "rdot"])
```
Comment (Akash-Maurya-0899) [inline gwnr/waveform/esigma_utils.py:793]: This is a bit dangerous because if the user does not enter a valid orbital variable name, they will not get any error now (via the code in `get_inspiral_esigma_modes()` function). I think an error is better than just mysteriously not getting anything for a wrongly typed variable name. The user should know which variables are available and which are not.

File: gwnr/waveform/esigma_utils.py [hunk 13]
```diff
@@ -468,83 +675,132 @@ def get_imr_esigma_modes(
...
-        f_Kerr = f_ISCO_spin(mass1, mass2, spin1z, spin2z)
-        # Schwarzschild ISCO frequency
-        f_Schwarz = 6.0**-1.5 / (mass1 + mass2) / lal.MTSUN_SI / lal.PI
-        f_mr_transition = min(f_Kerr, f_Schwarz)
-
-    if not return_orbital_params:
-        return_orbital_params = ["e"]
-        # Needed necessarily for hybridization error printing
     return_orbital_params = set(return_orbital_params)
+    return_orbital_params_user = set(return_orbital_params)
```
Comment (Akash-Maurya-0899) [inline gwnr/waveform/esigma_utils.py:800]: Did you mean to do `return_orbital_params_user = set(return_orbital_params_user)` here?

File: gwnr/waveform/hybridize.py [hunk 14]
```diff
@@ -313,9 +341,23 @@ def hybridize_modes(
     the one at the rightmost occurance in time) 
 
     """
-    t2_index_insp = find_last_value_location_in_series(
-        frq_insp[(el, em)], frq_attach + frq_width / 2
-    )
+    if hybridize_using_avg_orbital_frequency:
+        t2_index_insp = find_last_value_location_in_series(
+            inspiral_orbital_frequency, (frq_attach + frq_width / 2) / em
+        )
```
Comment (Akash-Maurya-0899) [inline gwnr/waveform/hybridize.py:347]: This function (`hybridize_modes()`) assumes `frq_attach` and `freq_width` should be in accordance with the `em` value of `mode_to_align_by`. Hence change this to
```
 t2_index_insp = find_last_value_location_in_series(
            inspiral_orbital_frequency, (frq_attach + frq_width / 2)
        )
```

**Review by Akash-Maurya-0899 — CHANGES_REQUESTED** [review]

File: gwnr/waveform/esigma_utils.py (hunk 13 above)
Comment (prayush) [reply gwnr/waveform/esigma_utils.py:800]: Not really, in fact this line is redundant and I've removed it. `return_orbital_params_user` basically tracks the parameters to be returned to user, excluding any params we might need for internal purposes. Move this line in the `if..elif..elif` clause above.

File: gwnr/waveform/esigma_utils.py (hunk 12 above)
Comment (prayush) [reply gwnr/waveform/esigma_utils.py:793]: This is a good point, but I think a warning will be sufficient instead of an error here. It would be good to return the primary objects of this function, ie the IMR modes, despite typos in the orbital variable names. I have added a warning print here to make this very very explicit.

File: gwnr/waveform/esigma_utils.py [hunk 15]
```diff
@@ -443,23 +446,227 @@ def get_inspiral_esigma_waveform(
...
+            # this could be optimized to not repeat integration
+            if abs(integrate.trapezoid(freq_[:idx], dx=delta_t_)) >= delta_phi_:
+                return idx
+
+
+def _get_transition_frequency_window(
+    orbital_phase,
+    orbital_freq,
+    delta_t,
+    f_mr_transition,  # dominant-mode GW frequency
```
[the remark this answers is not in the document]
Comment (prayush) [reply gwnr/waveform/esigma_utils.py:481]: thanks, fixed

File: gwnr/waveform/esigma_utils.py [hunk 16]
```diff
@@ -443,23 +446,227 @@ def get_inspiral_esigma_waveform(
...
+                orbital_phase, orbital_phase[transition_idx] - num_hyb_orbits * np.pi
+            )
+            window_end_idx = np.searchsorted(
+                orbital_phase, orbital_phase[transition_idx] + num_hyb_orbits * np.pi
+            )
+        f_window_mr_transition = 2 * np.min(
+            [
+                orbital_freq[window_end_idx] - f_mr_transition / 2.0,
+                f_mr_transition / 2.0 - orbital_freq[window_start_idx],
+            ]
```
[the remark this answers is not in the document]
Comment (prayush) [reply gwnr/waveform/esigma_utils.py:596]: thanks for catching this, fixed it.

File: gwnr/waveform/esigma_utils.py [hunk 17]
```diff
@@ -554,11 +810,22 @@ def get_imr_esigma_modes(
...
+        # These will be used for figuring out the hybridization window
+        if hybridize_using_avg_orbital_frequency:
+            return_orbital_params = return_orbital_params.union(
+                set(["x", "phi", "phidot"])
+            )
 
-    return_orbital_params = list(return_orbital_params)
+    _, mode_to_align_by_em = mode_to_align_by
+
+    # If the user does not provide the 22-mode frequency at which the inspiral
```
[the remark this answers is not in the document]
Comment (prayush) [reply gwnr/waveform/esigma_utils.py:821]: fixed here, and 2 lines below

File: gwnr/waveform/esigma_utils.py [hunk 18]
```diff
@@ -554,11 +810,22 @@ def get_imr_esigma_modes(
     # merger-ringdown, we switch schemes and hybridize over `num_hyb_orbits`
     # orbits instead.
     if f_window_mr_transition is None:
-        return_orbital_params = return_orbital_params.union(
-            set(["phi", "phidot"])
-        )  # These will be used for figuring out the hybridization window
+        # These will be used for figuring out the hybridization window
+        if hybridize_using_avg_orbital_frequency:
```
[the remark this answers is not in the document]
Comment (prayush) [reply gwnr/waveform/esigma_utils.py:814]: changed to:

```
        return_orbital_params = return_orbital_params.union(set(["phi", "phidot"]))
        if hybridize_using_avg_orbital_frequency:
            return_orbital_params = return_orbital_params.union(set(["x"]))
```

File: gwnr/waveform/esigma_utils.py [hunk 19]
```diff
@@ -443,23 +446,227 @@ def get_inspiral_esigma_waveform(
...
+            )
+            if window_start_idx is None:
+                if failsafe:
+                    window_start_idx = 0
+                else:
+                    raise RuntimeError(
+                        f"""Requested number of orbits to hybridize over not available
+in the waveform before the transition frequency. Either decrease the number of
+orbits to hybridize over (currently {num_hyb_orbits}) or decrease the
+inspiral-to-merger transition frequency. `window_start_idx` is None."""
```
[the remark this answers is not in the document]
Comment (prayush) [reply gwnr/waveform/esigma_utils.py:616]: fixed

File: gwnr/waveform/esigma_utils.py [hunk 20]
```diff
@@ -588,6 +559,27 @@ def _get_transition_frequency_window(
...
             )
+            # FAILSAFE mode behavior
+            if window_start_idx is None:
+                if failsafe:
+                    window_start_idx = 0
+                else:
+                    raise RuntimeError(
+                        f"""Requested number of orbits to hybridize over not available
+in the waveform before the transition frequency. Either decrease the number of
+orbits to hybridize over (currently {num_hyb_orbits}) or decrease the
```
[the remark this answers is not in the document]
Comment (prayush) [reply gwnr/waveform/esigma_utils.py:570]: fixed

File: gwnr/waveform/esigma_utils.py (hunk 1 above)
Comment (prayush) [reply gwnr/waveform/esigma_utils.py:643]: can we resolve this @Akash-Maurya-0899 ?

File: gwnr/waveform/esigma_utils.py [hunk 21]
```diff
@@ -443,23 +446,227 @@ def get_inspiral_esigma_waveform(
...
+    if direction == "backward":
+        for idx in range(len(freq_) - 2, 0, -1):
+            # this could be optimized to not repeat integration
+            if abs(integrate.trapezoid(freq_[idx:], dx=delta_t_)) >= delta_phi_:
+                return idx
+    elif direction == "forward":
+        for idx in range(2, len(freq_)):
+            # this could be optimized to not repeat integration
+            if abs(integrate.trapezoid(freq_[:idx], dx=delta_t_)) >= delta_phi_:
+                return idx
```
[the remark this answers is not in the document]
Comment (prayush) [reply gwnr/waveform/esigma_utils.py:474]: the last iteration in both your suggestions will throw errors, when trying to access

```
freq_[:len(freq_) + 1]
```
so the upper limits of index should not cross `len(freq_)`.

Now, as to whether to return `idx` or `idx-1`, I think we should be consistent with the `direction="backward"` case where the returned index corresponds to the sample where `delta_phi` radians of phase has already elapsed. IE we slightly have overshot. If we return `idx-1` (both of your suggestions effectively), then we would be returning the index where `delta_phi` radians of phase is just about to elapse, but has not yet.

In short, I think the current code is just fine.

File: gwnr/waveform/esigma_utils.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (prayush) [reply gwnr/waveform/esigma_utils.py:1008]: done

File: gwnr/waveform/esigma_utils.py [hunk 22]
```diff
@@ -530,52 +528,25 @@ def _get_transition_frequency_window(
...
 
-        if verbose > 4:
-            print(
-                f"""The transition is to happen in a window from
-                frequencies: [{orbital_freq[window_start_idx]}, {orbital_freq[transition_idx]}]Hz,
-                between indices: [{window_start_idx}, {transition_idx}]"""
+        tmp_ = [0]
+        for idx in range(2, len(orbital_freq)):
+            tmp_.append(integrate.trapezoid(orbital_freq[:idx], dx=delta_t))
+        orbital_phase = np.asarray(tmp_)
```
[the remark this answers is not in the document]
Comment (prayush) [reply gwnr/waveform/esigma_utils.py]: basic tests suggest that this would produce the same phase

**Comment by prayush** [comment]
@Akash-Maurya-0899 good to go again

File: gwnr/waveform/esigma_utils.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (prayush) [reply gwnr/waveform/esigma_utils.py:473]: Thanks for the correction, I'm going with the second option.
